internal/tasks/publishertask_test.go

```diff
@@ -0,0 +1,246 @@
+	"github.com/stretchr/testify/assert"
+)
+
+// GetDummyTopicModel to export dummy topic model
+func GetDummyTopicModel() []*topic.Model {
+	return []*topic.Model{
+		{
+			Name:               "projects/test-project/topics/test",
+			NumPartitions:      2,
+			ExtractedTopicName: "test",
+			ExtractedProjectID: "test-project",
+			Labels:             map[string]string{},
+		},
+	}
+}
+
+func setupPT(t *testing.T) (
+	ctx context.Context,
+	ctrl *gomock.Controller,
+	cancel context.CancelFunc,
+	registryMock *mocks.MockIRegistry,
+	topicCoreMock *mocks4.MockICore,
+) {
+	ctx, cancel = context.WithCancel(context.Background())
+	ctrl = gomock.NewController(t)
+
+	registryMock = mocks.NewMockIRegistry(ctrl)
+	topicCoreMock = mocks4.NewMockICore(ctrl)
+	return
+}
+
+func TestNewPublisherTask(t *testing.T) {
+	type args struct {
+		id        string
+		registry  registry.IRegistry
+		topicCore topic.ICore
+		options   []Option
+	}
+	_, ctrl, cancel, registryMock, topicCoreMock := setupPT(t)
+	defer ctrl.Finish()
+	defer cancel()
+
+	task := &PublisherTask{
+		id:             uuid.New().String(),
+		registry:       registryMock,
+		topicCore:      topicCoreMock,
+		topicWatchData: make(chan *struct{}),
+	}
+
+	tests := []struct {
+		name    string
+		args    args
+		want    ITask
+		wantErr bool
+	}{
+		{
+			name: "Creating new publisher task successfully",
+			args: args{
+				id:        uuid.New().String(),
+				registry:  registryMock,
+				topicCore: topicCoreMock,
+				options:   []Option{},
+			},
+			want:    task,
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			_, err := NewPublisherTask(tt.args.id, tt.args.registry, tt.args.topicCore, tt.args.options...)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("NewPublisherTask() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+		})
+	}
+}
+
+func TestPublisherTask_Run(t *testing.T) {
+	ctx, ctrl, cancel, registryMock, topicCoreMock := setupPT(t)
+	defer ctrl.Finish()
+	defer cancel()
+	watcherMock := mocks.NewMockIWatcher(ctrl)
+
+	workerID := uuid.New().String()
+	task, err := NewPublisherTask(
+		workerID,
+		registryMock,
+		topicCoreMock)
+	assert.Nil(t, err)
+
+	doneCh := make(chan struct{})
+
+	// mock registry
+	registryMock.EXPECT().Watch(gomock.AssignableToTypeOf(ctx), gomock.Any()).Return(watcherMock, nil).AnyTimes()
+
+	// mock watcher
+	watcherMock.EXPECT().StartWatch().Do(func() {
+		<-ctx.Done()
+	}).Return(nil)
+	watcherMock.EXPECT().StopWatch()
+
+	dummyTopicModels := GetDummyTopicModel()
+	// mock Topic Get
+	topicCoreMock.EXPECT().List(gomock.AssignableToTypeOf(ctx), "topics/").Return(
+		dummyTopicModels, nil).AnyTimes()
+
+	go func() {
+		err = task.Run(ctx)
+		assert.Equal(t, err, context.Canceled)
+		close(doneCh)
+	}()
+
+	// test signals on channels
+	task.(*PublisherTask).topicWatchData <- &struct{}{}
+
+	// test nils
+	task.(*PublisherTask).topicWatchData <- nil
+
+	cancel()
+	<-doneCh
```

Comment:
what is the purpose of this..? Start watch will always cancel the context, and hence we will get err on Run().
Where exactly are we consuming from these channels..?

Reply:
@sakshi-goyal-razorpay the purpose of this is to stopWatch at the end, if we remove this stopwatch will not be executed.